homeassistant/components/opensky/sensor.py

```diff
@@ -115,58 +130,119 @@ def state(self):
         """Return the state of the sensor."""
         return self._state
 
-    def _handle_boundary(self, flights, event, metadata):
-        """Handle flights crossing region boundary."""
-        for flight in flights:
-            if flight in metadata:
-                altitude = metadata[flight].get(ATTR_ALTITUDE)
-            else:
-                # Assume Flight has landed if missing.
-                altitude = 0
-
-            data = {
-                ATTR_CALLSIGN: flight,
-                ATTR_ALTITUDE: altitude,
-                ATTR_SENSOR: self._name,
-            }
-            self._hass.bus.fire(event, data)
-
-    def update(self):
+    def _calc_bbox(self):
+        """
+        Calculate the rectangular bounding box for a circle on Earth's surface.
+
+        Credit:
+          - Based on code written in Java by Jan Matuschek:
+            http://janmatuschek.de/LatitudeLongitudeBoundingCoordinates#Java
+          - Later ported to Python by Jeremy Fein:
+            https://github.com/jfein/PyGeoTools
+        License: https://creativecommons.org/licenses/by/3.0/
+        Changes: Adapted for Home Assistant's opensky sensor
+        """
+        rad_lat = math.radians(self._latitude)
+        rad_lon = math.radians(self._longitude)
+
+        # angular distance in radians on a great circle
+        rad_dist = self._radius / util_location.AXIS_A
+
+        min_lat = rad_lat - rad_dist
+        max_lat = rad_lat + rad_dist
+
+        if min_lat > MIN_LAT and max_lat < MAX_LAT:
+            delta_lon = math.asin(math.sin(rad_dist) / math.cos(rad_lat))
+
+            min_lon = rad_lon - delta_lon
+            if min_lon < MIN_LON:
+                min_lon += 2 * math.pi
+
+            max_lon = rad_lon + delta_lon
+            if max_lon > MAX_LON:
+                max_lon -= 2 * math.pi
+        # a pole is within the distance
+        else:
+            min_lat = max(min_lat, MIN_LAT)
+            max_lat = min(max_lat, MAX_LAT)
+            min_lon = MIN_LON
+            max_lon = MAX_LON
+
+        self._params["lamin"] = str(round(math.degrees(min_lat), 4))
+        self._params["lomin"] = str(round(math.degrees(min_lon), 4))
+        self._params["lamax"] = str(round(math.degrees(max_lat), 4))
+        self._params["lomax"] = str(round(math.degrees(max_lon), 4))
+
+    async def async_update(self):
         """Update device state."""
         currently_tracked = set()
         flight_metadata = {}
-        states = self._session.get(OPENSKY_API_URL).json().get(ATTR_STATES)
-        for state in states:
-            flight = dict(zip(OPENSKY_API_FIELDS, state))
-            callsign = flight[ATTR_CALLSIGN].strip()
-            if callsign != "":
-                flight_metadata[callsign] = flight
-            else:
+
+        session = async_get_clientsession(self._hass)
+        try:
+            async with session.get(OPENSKY_API_URL, params=self._params) as resp:
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices